### PR TITLE
[DOCS] Tell the bot's story to children

### PR DESCRIPTION
**Pull request by jpic**
For posterity, and to help fellow hackers trying to make some sense out of this
great FMS example !

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -53,6 +53,36 @@ Now write a ``jenkins.yml`` file and open a PR::
 Many instructions are available. Just ask the bot by commenting ``jenkins:
 help`` in an open PR!
 
+The bot's story
+===============
+
+The bot executes the extensions listed in setup.py's entry point
+'jenkins_epo.bot.extensions'. Each entry point defines a stage number which is
+used by the bot to decide in which order to execute extensions.
```
Comment (bersace) [inline README.rst]: Une petite mention de `list-extensions` qui permet de visualiser le pipeline du bot serat pas mal, non ?

File: README.rst [hunk 2]
```diff
@@ -53,6 +53,36 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+
+The ``bot.run()`` asyncio coroutine invoked by the ``jenkins-epo bot`` command
+creates a payload in ``self.current`` that will be passed in each extension's
```
Comment (bersace) [inline README.rst]: Je dirais plutôt un contexte plutôt que payload. C'est un peu comme un contexte de requête en Django.

File: README.rst [hunk 3]
```diff
@@ -53,6 +53,36 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+method call. Note that the ``LOOP=1`` environment variable will make the
+``jenkins-epo bot`` command to loop forever over calls of ``bot.run()``.
+
+First step of ``bot.run()`` is executing the ``ext.begin()`` method of each
+extension. This method should encapsulate any of the extension's variable
+initialization and eventually kick the payload out ouf ``bot.run()`` by raising
```
Comment (bersace) [inline README.rst]: s/ouf/of/

Comment (bersace) [inline README.rst]: > kick the payload out ouf ``bot.run()``

Je dirais plutôt `abort the pipeline`.

File: README.rst [hunk 4]
```diff
@@ -53,6 +53,36 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+loop.
+
+Then, ``bot.run()`` executes the ``ext.process_instruction(instruction)``
+extension method for each instruction parsed from the github comments by the
+bot. You'll note from examples such as the OPM extension that the bot gives
+instructions to itself in HTML comments when posting user feedback. This
+demonstrates how to maintain a state for an extension on a pull request.
+
+Finnaly, the ``ext.run()`` extension coroutine is called, again with the
+opportunity to raise ``SkipHead`` and directly go to the end of ``bot.run()``
```
Comment (bersace) [inline README.rst]: Je ferai un seul paragraphe pour `SkipHead`: *Extension can abort the pipeline either in `begin()` or `run()` by raising `SkipHead` ad-hoc exception.* etc.

**Comment by jpic** [comment]
Suggestions taken care of, please review changes :dancer:

File: README.rst [hunk 5]
```diff
@@ -53,6 +53,35 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+The bot's story
+===============
+
+The bot executes the extensions listed in setup.py's entry point
+'jenkins_epo.bot.extensions'. Each entry point defines a stage number which is
+used by the bot to decide in which order to execute extensions. The
+``jenkins-epo list-extensions`` command outputs the pipeline.
+
+The ``bot.run()`` asyncio coroutine invoked by the ``jenkins-epo bot`` command
+creates a payload in ``self.current`` available in each extension's method
```
Comment (bersace) [inline README.rst]: s/payload/context/

Pour moi, payload, c'est le retour d'un appel à l'API GitHub ^^

File: README.rst [hunk 6]
```diff
@@ -53,6 +53,35 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+call. Note that the ``LOOP=1`` environment variable will make the ``jenkins-epo
+bot`` command to loop forever over calls of ``bot.run()``.
+
+First step of ``bot.run()`` is calling the ``ext.begin()`` method of each
+extension. This method should encapsulate any of the extension's variable
```
Comment (bersace) [inline README.rst]: variable**s** ?

File: README.rst [hunk 7]
```diff
@@ -53,6 +53,35 @@ Now write a ``jenkins.yml`` file and open a PR::
 Many instructions are available. Just ask the bot by commenting ``jenkins:
 help`` in an open PR!
 
+The bot's story
+===============
+
+The bot executes the extensions listed in setup.py's entry point
```
Comment (bersace) [inline README.rst]: Par exemple:

*The bot loads extensions from the setuptools entrypoint `jenkins_epo.bot.extensions`. Each extension has a `stage` key to order extensions execution. The `jenkins-epo list-extensions` command outputs the pipeline.*

File: README.rst [hunk 8]
```diff
@@ -53,6 +53,35 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+'jenkins_epo.bot.extensions'. Each entry point defines a stage number which is
+used by the bot to decide in which order to execute extensions. The
+``jenkins-epo list-extensions`` command outputs the pipeline.
+
+The ``bot.run()`` asyncio coroutine invoked by the ``jenkins-epo bot`` command
```
Comment (bersace) [inline README.rst]: Ici:

*`jenkins-epo bot` creates an isolated context for each branch or PR to test in extension's `self.current`.*

File: README.rst [hunk 9]
```diff
@@ -53,6 +53,35 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+extension. This method should encapsulate any of the extension's variable
+initialization.
+
+Then, ``bot.run()`` executes the ``ext.process_instruction(instruction)``
+extension method for each instruction parsed from the github comments by the
+bot. You'll note from examples such as the OPM extension that the bot gives
+instructions to itself in HTML comments when posting user feedback. This
+demonstrates how to maintain a state for an extension on a pull request.
+
+Finnaly, the ``ext.run()`` extension coroutine is called for each extension.
```
Comment (bersace) [inline README.rst]: Ici:

*Finnaly, `jenkins-epo bot` calls each ``ext.run()`` coroutine. This is where the extension works with the different REST APIs.*

File: README.rst [hunk 10]
```diff
@@ -53,6 +53,35 @@ Now write a ``jenkins.yml`` file and open a PR::
...
+This is where the extension is executed and works with the different APIs.
+
+A special ``SkipHead`` exception may be raised by the extension's ``ext.begin()``
```
Comment (bersace) [inline README.rst]: Ici:

*`ext.begin()` or `ext.run()` can tell the bot to abort the pipeline on a branch by raising the special exception `SkipHead`.*

**Pull request merged**